internal/ls/folding.go

```diff
@@ -0,0 +1,541 @@
+		}
+	}
+
+	span := getOutliningSpanForNode(n, sourceFile, l)
+	if span != nil {
+		foldingRange = append(foldingRange, span)
+	}
+
+	depthRemaining--
+	if ast.IsCallExpression(n) {
+		depthRemaining++
+		expressionNodes := visitNode(n.Expression(), depthRemaining, sourceFile, l)
+		if expressionNodes != nil {
+			foldingRange = append(foldingRange, expressionNodes...)
+		}
+		depthRemaining--
+		for _, arg := range n.Arguments() {
+			if arg != nil {
+				foldingRange = append(foldingRange, visitNode(arg, depthRemaining, sourceFile, l)...)
+			}
+		}
+		typeArguments := n.TypeArguments()
+		for _, typeArg := range typeArguments {
+			if typeArg != nil {
+				foldingRange = append(foldingRange, visitNode(typeArg, depthRemaining, sourceFile, l)...)
+			}
+		}
+	} else if ast.IsIfStatement(n) && n.AsIfStatement().ElseStatement != nil && ast.IsIfStatement(n.AsIfStatement().ElseStatement) {
+		// Consider an 'else if' to be on the same depth as the 'if'.
+		ifStatement := n.AsIfStatement()
+		expressionNodes := visitNode(n.Expression(), depthRemaining, sourceFile, l)
+		if expressionNodes != nil {
+			foldingRange = append(foldingRange, expressionNodes...)
+		}
+		thenNode := visitNode(ifStatement.ThenStatement, depthRemaining, sourceFile, l)
+		if thenNode != nil {
+			foldingRange = append(foldingRange, thenNode...)
+		}
+		depthRemaining++
+		elseNode := visitNode(ifStatement.ElseStatement, depthRemaining, sourceFile, l)
+		if elseNode != nil {
+			foldingRange = append(foldingRange, elseNode...)
+		}
+		depthRemaining--
+	} else {
+		visit := func(node *ast.Node) bool {
+			childNode := visitNode(node, depthRemaining, sourceFile, l)
+			if childNode != nil {
+				foldingRange = append(foldingRange, childNode...)
+			}
+			return false
+		}
+		n.ForEachChild(visit)
+	}
+	depthRemaining++
+	return foldingRange
+}
+
+func addOutliningForLeadingCommentsForNode(n *ast.Node, sourceFile *ast.SourceFile, l *LanguageService) []*lsproto.FoldingRange {
+	if ast.IsJsxText(n) {
+		return nil
+	}
+	return addOutliningForLeadingCommentsForPos(n.Pos(), sourceFile, l)
+}
+
+func addOutliningForLeadingCommentsForPos(pos int, sourceFile *ast.SourceFile, l *LanguageService) []*lsproto.FoldingRange {
+	p := &printer.EmitContext{}
+	comments := scanner.GetLeadingCommentRanges(&printer.NewNodeFactory(p).NodeFactory, sourceFile.Text(), pos)
+
+	var foldingRange []*lsproto.FoldingRange
+	firstSingleLineCommentStart := -1
+	lastSingleLineCommentEnd := -1
+	singleLineCommentCount := 0
+	foldingRangeKindComment := lsproto.FoldingRangeKindComment
+
+	combineAndAddMultipleSingleLineComments := func() *lsproto.FoldingRange {
+		// Only outline spans of two or more consecutive single line comments
+		if singleLineCommentCount > 1 {
+			return createFoldingRangeFromBounds(firstSingleLineCommentStart, lastSingleLineCommentEnd, foldingRangeKindComment, sourceFile, l)
+		}
+		return nil
+	}
+
+	sourceText := sourceFile.Text()
+	comments(func(comment ast.CommentRange) bool {
```

Comment:
Not that you can't, but why was this written this way instead of `for comment := range GetLeadingCommentRanges`?